### PR TITLE
Add OMEGA Broadcast functions

### PR DESCRIPTION
**Pull request by grnydawn**
This PR adds OMEGA broadcasting functions with documentation.
* implements blocking broadcast functions
* adds testings for the functions
* adds user-guide and developer-guide

Checklist
* [X] Documentation:
  * [X] Design document has been generated and added to the docs
  * [X] User's Guide has been updated
  * [X] Developer's Guide has been updated
  * [ ] Documentation has been [built locally](https://e3sm-project.github.io/Omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [X] Testing
  * [ ] A comment in the PR documents testing used to verify the changes including any tests that are added/modified/impacted.
  * [X] CTest unit tests for new features have been added per the approved design. 
  * [ ] Polaris tests for new features have been added per the approved design (and included in a test suite)
  * [] Unit tests have passed. Please provide a relevant CDash build entry for verification.
  * [ ] Polaris test suite has passed
  * [ ] Performance related PRs: Please include a relevant PACE experiment link documenting performance before and after.
* [ ] Stealth Features
  * [ ] If any stealth features are included in the PR, please confirm that they have been documented.

NOTE: Non-blocking broadcast functions are NOT included in this PR.

### CODE REVIEW

**Comment by grnydawn** [comment]
@philipwjones , I have made modifications to two lines of the Broadcast design document, where the MachEnv function argument has been updated to a pointer variable.

File: components/omega/src/infra/Broadcast.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#include "Broadcast.h"
+
+namespace OMEGA {
+
+//------------------------------------------------------------------------------
+// Broadcast I4 scalar value
+int Broadcast(I4 &value, const MachEnv *InEnv, const int rankBcast) {
```
Comment (philipwjones) [inline components/omega/src/infra/Broadcast.cpp:19]: Following the style guide, all vars start with a capital letter (I still have trouble with this too as evidenced by the design doc). So Value, RankBcast and Retval/RetVal throughout.

File: components/omega/src/infra/Broadcast.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+//===----------------------------------------------------------------------===//
+
+#include "DataTypes.h"
+#include "MachEnv.h"
+#include "mpi.h"
+
+namespace OMEGA {
+
+// blocking broadcast scalar
+int Broadcast(I4 &value, const MachEnv *InEnv = MachEnv::getDefaultEnv(),
```
Comment (philipwjones) [inline components/omega/src/infra/Broadcast.h:20]: One big change here. We allow the MachEnv to define a MasterTask other than 0, so in all these interfaces, we want to set the argument RankBcast to an invalid value and extract MasterTask from the MachEnv.

**Review by philipwjones — CHANGES_REQUESTED** [review]
Other than the changes noted, this looks good and passes tests on Chrysalis. I'll try building the docs and re-test when the rank and style changes are done.

**Comment by philipwjones** [comment]
A few more changes after building the docs. Remember to add the Broadcast docs to the top-level index.md file.

Also, could you make all the headings other than the main broadcast heading a level lower than the main heading. Right now, many of the subsections are showing up as top-level sections in the contents.

**Comment by philipwjones** [comment]
One last (?) thing - I just realized these are in the infra directory. Can you please move them over to the base directory? That's where much of the MPI-related infrastructure should sit. The infra directory should be for utilities that are less dependent on the MPI model (eg Config, Logging, TimeManager, etc). That means also moving the docs, etc to match the change in location. Thanks.

File: components/omega/src/infra/Broadcast.cpp (hunk 1 above)
Comment (grnydawn) [reply components/omega/src/infra/Broadcast.cpp]: renamed variables as suggested.

File: components/omega/src/infra/Broadcast.h (hunk 2 above)
Comment (grnydawn) [reply components/omega/src/infra/Broadcast.h]: Thanks a lot for pointing this. Now, if user does not specify RankBcast value(default value is set to an invalid number of "-1"), the module get it using getMasterTask().

**Comment by grnydawn** [comment]
@philipwjones Thanks a lot for the suggestions. I think I have applied all of your suggestions, and please review if the changes match with your thought.

**Review by philipwjones — APPROVED** [review]
Thanks for making the changes. I have confirmed unit tests pass on both Chrysalis (intel) and Frontier (crayclangcpu). I also built the documentation and it looks fine.

**Review by xylar — APPROVED** [review]
I'm approving just based on a visual inspection of the docs.  I don't have time at the moment to do a code review, sorry about that.  But I think @philipwjones' review is sufficient.

**Pull request merged**